cachi2/core/package_managers/yarn/locators.py

```diff
@@ -0,0 +1,236 @@
+from typing import NamedTuple, Protocol
+
+from cachi2.core.errors import PackageRejected
+
+
+class Locator(Protocol):
```

Comment:
I'm mostly fiddling with Python ~interfaces~ protocols here, I'm not sure I like them here yet.

Reply:
I decided to leave the purl generation out of the locator models though, since they would rely on external attributes (actual name from package.json, main module version for local files/worskpaces).

Another reason is that I'm seeing these as mostly parsers for the string locators, so they shouldn't have knowledge about purl generation.

Reply:
I'm not sure if the Locator Protocol is useful at all

`name` is only useful for one particular Locator type, `protocol` probably isn't useful for any and `from_string` doesn't need to be part of it since parse_locator will work either way

I think a good old Union would the best. We have a fixed set of known locators, so the type might as well reflect that (as opposed to a protocol which can have infinitely many implementers, so you can't do exhaustiveness checking on it)